### PR TITLE
Javascript dependent styling

### PR DESCRIPTION
**Pull request by alexkilgour**
Add a section in `practices` about how to bootstrap Javascript dependent styles, then link to the information when we talk about it in CSS house style. Mostly cribbed from something @sonniesedge wrote (🙏 ) but removing specific references to components and the design system.

(Also fixes some dead links from the Javascript house style TOC)

### CODE REVIEW

File: practices/javascript-styling.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# Javascript dependent styling
+
+A class name of `.js` on the document root element (normally `html`) will be used to bootstrap JS-only visual styles. We do this to prevent the visual flicker and browser reflow that may occur when JavaScript applies CSS styles to a component after the page has loaded.
```
Comment (jpw) [inline practices/javascript-styling.md:3]:
```suggestion
A class name of `.js` on the document root element (normally `html`) is used to target JS-only visual styles via CSS. We do this to minimise any visual flickering and browser reflows that may occur when JavaScript applies CSS styles to a component after the page has loaded.
```

I _think_ my suggestion is a little more accurate. But I am not clear on what the second sentence is saying -- it doesn't I think prevent reflows etc? Or is this in comparison to another, unmentioned technique?

File: practices/javascript-styling.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+## Scripts
+
+The `.js` class must be added via a micro (but blocking) script, placed as far up in the `<head>` of the document as possible.
```
Comment (jpw) [inline practices/javascript-styling.md:7]:
```suggestion
The `.js` class must be added via a micro (but blocking) script, placed as far up in the `<head>` of the document as possible.

The blocking nature of the script prevents any visual flickering or browser reflows, unlike other techniques where JavaScript may cause CSS styles to be applied after the page has rendered.
```

Is this what we are driving at?

Comment (alexkilgour) [reply practices/javascript-styling.md:7]: Yep, exactly this (at least this is also my assumption, @sonniesedge wrote it!)

File: practices/javascript-styling.md (hunk 1 above)
Comment (sonniesedge) [reply practices/javascript-styling.md:3]: Maybe "JS-dependent" rather than "JS-only".

File: practices/javascript-styling.md (hunk 2 above)
Comment (sonniesedge) [reply practices/javascript-styling.md:7]: Yup, that's exactly it. Although I can't remember writing this. 

If it's a success I shall claim credit. If not, Alex wrote it and blamed me.

File: practices/javascript-styling.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+```html
+<script>
+    // JS Detection Script
+    (function(e){var t=e.documentElement,n=e.implementation;t.className='js';})(document)
+</script>
+```
+
+## Styles
+
+All UI elements MUST be usable and visually-acceptable when Javascript is not present. Enhancements can be made when Javascript is present by nesting those styles underneath the `.js` class.
```
Comment (sonniesedge) [inline practices/javascript-styling.md:18]: Maybe add at the end "As always with [Progressive Enhancement](progressive-enhancement.md), try building the core version first, delivering it quickly to stakeholders. Then build and deliver the enhanced version in a subsequent iteration."

File: css/house-style.md [hunk 4]
```diff
@@ -152,14 +152,14 @@ table {
 **Non-js fallbacks/JS only styles**. We do this:
 ```scss
 .c-class {
+	display: block;
+
     .js & {
         display: none;
```
Comment (sonniesedge) [inline css/house-style.md:158]: Maybe a different enhanced example here? It looks odd that a JS enhancement is actually hiding something. `display: flex` instead?

**Comment by sonniesedge** [comment]
Excellent stuff.

File: practices/javascript-styling.md (hunk 1 above)
Comment (jpw) [reply practices/javascript-styling.md:3]: nice, done ty

File: practices/javascript-styling.md [hunk 5]
```diff
@@ -19,8 +19,10 @@ The blocking nature of the script prevents any visual flickering or browser refl
 
 All UI elements MUST be usable and visually-acceptable when Javascript is not present. Enhancements can be made when Javascript is present by nesting those styles underneath the `.js` class.
 
+Following the principles of [Progressive Enhancement](./progressive-enhancement.md), consider buiding and releasing your core unenhanced version first. Then build and deliver the javascript enhanced version in a subsequent iteration.
```
Comment (jpw) [inline practices/javascript-styling.md:22]:
```suggestion
Following the principles of [Progressive Enhancement](./progressive-enhancement.md), consider building and releasing your core unenhanced version first. Then build and deliver the Javascript enhanced version in a subsequent iteration.
```

**Comment by alexkilgour** [comment]
> Thanks very much Alex, I think this is now good to go?

No, thank YOU Jon 😄

**Pull request merged**